src/io/LivermoreParamsReader.cc

```diff
@@ -0,0 +1,210 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file LivermoreParamsReader.cc
+//---------------------------------------------------------------------------//
+#include "LivermoreParamsReader.hh"
+
+#include <fstream>
+#include <sstream>
+#include "comm/Logger.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the reader, optionally supplying the path to the directory
+ * containing the data. If no path is specified, the environment variable
+ * G4LEDATA is used to locate the data.
+ */
+LivermoreParamsReader::LivermoreParamsReader(const char* path)
+{
+    if (path)
+    {
+        std::ostringstream os;
+        os << path << "/";
+        path_ = os.str();
+    }
+    else
+    {
+        const char* env_var = std::getenv("G4LEDATA");
+        if (!env_var)
+        {
+            CELER_LOG(error) << "Environment variable G4LEDATA is not defined";
+        }
+        std::ostringstream os;
+        os << env_var << "/livermore/phot_epics2014/";
+        path_ = os.str();
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Read the data for the given elements.
+ */
+LivermoreParamsReader::result_type
+LivermoreParamsReader::operator()(ElementDefId el_id, int atomic_number)
+{
+    REQUIRE(atomic_number > 0 && atomic_number < 101);
+
+    constexpr real_type barn = 1.e-24 * units::centimeter * units::centimeter;
+
+    result_type result;
+    result.el_id = el_id;
+
+    // Read photoelectric effect total cross section above K-shell energy but
+    // below energy limit for parameterization
+    std::string Z        = std::to_string(atomic_number);
+    std::string filename = path_ + "pe-cs-" + Z + ".dat";
+    if (file_exists(filename))
+    {
+        std::ifstream infile(filename);
+        CHECK(infile.is_open());
+
+        // Set the physics vector type and the data type
+        result.xs_high.vector_type = ImportPhysicsVectorType::free;
+
+        // Read tabulated energies and cross sections
+        real_type energy_min, energy_max;
+        size_type size;
+        infile >> energy_min >> energy_max >> size >> size;
```

Comment:
After these kind of streaming operators, it's important to check that the file is valid, otherwise the data might be uninitialized, and the following resize call will probably crash. I would initialize these variables to zero.

Also, is the `size` supposed to be there twice?

Reply:
Good call, thanks! And yes, the size is there twice 😄